File: rxbinding/src/main/java/com/jakewharton/rxbinding2/widget/RadioGroupCheckedChangeObservable.java
package com.jakewharton.rxbinding2.widget;

import android.widget.RadioGroup;
import android.widget.RadioGroup.OnCheckedChangeListener;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class RadioGroupCheckedChangeObservable extends InitialValueObservable<Integer> {
  private final RadioGroup view;

  RadioGroupCheckedChangeObservable(RadioGroup view) {
    this.view = view;
  }

  @Override protected void subscribeListener(Observer<? super Integer> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    view.setOnCheckedChangeListener(listener);
    observer.onSubscribe(listener);
  }

  @Override protected Integer getInitialValue() {
    return view.getCheckedRadioButtonId();
  }

  static final class Listener extends MainThreadDisposable implements OnCheckedChangeListener {
    private final RadioGroup view;
    private final Observer<? super Integer> observer;
    private int lastChecked = -1;

    Listener(RadioGroup view, Observer<? super Integer> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onCheckedChanged(RadioGroup radioGroup, int checkedId) {
      if (!isDisposed() && checkedId != lastChecked) {
        lastChecked = checkedId;
        observer.onNext(checkedId);
      }
    }

    @Override protected void onDispose() {
      view.setOnCheckedChangeListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding4/widget/RadioGroupCheckedChangeObservable.kt
@file:JvmName("RxRadioGroup")
@file:JvmMultifileClass

package com.jakewharton.rxbinding4.widget

import android.widget.RadioGroup
import android.widget.RadioGroup.OnCheckedChangeListener
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding4.InitialValueObservable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding4.internal.checkMainThread

/**
 * Create an observable of the checked view ID changes in `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun RadioGroup.checkedChanges(): InitialValueObservable<Int> {
  return RadioGroupCheckedChangeObservable(this)
}

private class RadioGroupCheckedChangeObservable(
  private val view: RadioGroup
) : InitialValueObservable<Int>() {

  override fun subscribeListener(observer: Observer<in Int>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    view.setOnCheckedChangeListener(listener)
    observer.onSubscribe(listener)
  }

  override val initialValue get() = view.checkedRadioButtonId

  private class Listener(
    private val view: RadioGroup,
    private val observer: Observer<in Int>
  ) : MainThreadDisposable(), OnCheckedChangeListener {
    private var lastChecked = -1

    override fun onCheckedChanged(radioGroup: RadioGroup, checkedId: Int) {
      if (!isDisposed && checkedId != lastChecked) {
        lastChecked = checkedId
        observer.onNext(checkedId)
      }
    }

    override fun onDispose() {
      view.setOnCheckedChangeListener(null)
    }
  }
}
